nonce/nonce_test.go

```diff
@@ -118,3 +123,80 @@ func BenchmarkNonces(b *testing.B) {
 		}
 	})
 }
+
+func TestNoncePrefixing(t *testing.T) {
+	ns, err := NewNonceService(metrics.NewNoopScope(), 0, "zinc")
+	test.AssertNotError(t, err, "Could not create nonce service")
+
+	n, err := ns.Nonce()
+	test.AssertNotError(t, err, "Could not create nonce")
+	test.Assert(t, ns.Valid(n), "Valid nonce rejected")
+
+	n, err = ns.Nonce()
+	test.AssertNotError(t, err, "Could not create nonce")
+	n = n[1:]
+	test.Assert(t, !ns.Valid(n), "Valid nonce with incorrect prefix accepted")
+
+	n, err = ns.Nonce()
+	test.AssertNotError(t, err, "Could not create nonce")
+	test.Assert(t, !ns.Valid(n[6:]), "Valid nonce without prefix accepted")
+}
+
+type malleableNonceClient struct {
+	redeem func(ctx context.Context, in *noncepb.NonceMessage, opts ...grpc.CallOption) (*noncepb.ValidMessage, error)
+}
+
+func (mnc *malleableNonceClient) Redeem(ctx context.Context, in *noncepb.NonceMessage, opts ...grpc.CallOption) (*noncepb.ValidMessage, error) {
+	return mnc.redeem(ctx, in, opts...)
+}
+
+func (mnc *malleableNonceClient) Nonce(ctx context.Context, in *corepb.Empty, opts ...grpc.CallOption) (*noncepb.NonceMessage, error) {
+	return nil, errors.New("unimplemented")
+}
+
+func TestRemoteRedeem(t *testing.T) {
```

Comment:
This test body is pretty dense. Can you add some high-level comments about what is being tested/expected to happen?